packages/aws-cdk/lib/api/util/sdk.ts

```diff
@@ -134,59 +122,81 @@ export class SDK implements ISDK {
     return this.service(AWS.ECR, account, region, mode);
   }
 
-  public async defaultRegion(): Promise<string | undefined> {
-    return await getCLICompatibleDefaultRegionGetter(this.profile)();
+  /**
+   * Obtain an SDK initialized by assuming a role
+   *
+   * WARNING: This does not refresh any credentials, because I don't understand
+   * the applicable building blocks of the JS SDK well enough. There is a
+   * discrepancy between TemporaryCredentials and CredentialProviders that I
+   * don't fully grok.
+   */
+  public async assumeRole(arn: string, region: string): Promise<ISDK> {
+    debug(`Assuming role '${arn}'`);
+    const sts = await this.service(AWS.STS, (await this.defaultAccount())?.accountId, region, Mode.ForWriting);
+    const response = await sts.assumeRole({
```

Comment:
I believe the correct way is to use `ChainableTemporaryCredentials` instead of calling assume role directly.

Also, shouldn't this reused with the `cdk-assets` codebase somehow?

Reply:
Yeah, there is some logic that gets shared between `cdk-assets` and `aws-cdk`. Some typical SDK operations, authentication, but also the substitution of `${AWS::Region}` etc placeholders in Role ARNs (both `assetPublishRoleArn` and `cloudFormationExecutionRoleArn` can contain placeholders but they're interpreted by different tools).

We would need to set up yet another shared lib to share that logic though, so for now it's (somewhat) duplicated between them.